homeassistant/components/rituals_perfume_genie/switch.py

```diff
@@ -1,104 +1,77 @@
+from .entity import DiffuserEntity
 
-_LOGGER = logging.getLogger(__name__)
-
-SCAN_INTERVAL = timedelta(seconds=30)
+STATUS = "status"
+FAN = "fanc"
+SPEED = "speedc"
+ROOM = "roomc"
 
 ON_STATE = "1"
 AVAILABLE_STATE = 1
 
-MANUFACTURER = "Rituals Cosmetics"
-MODEL = "Diffuser"
-ICON = "mdi:fan"
-
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
     """Set up the diffuser switch."""
-    account = hass.data[DOMAIN][config_entry.entry_id]
-    diffusers = await account.get_devices()
-
+    diffusers = hass.data[DOMAIN][config_entry.entry_id][DEVICES]
+    coordinators = hass.data[DOMAIN][config_entry.entry_id][COORDINATORS]
     entities = []
-    for diffuser in diffusers:
-        entities.append(DiffuserSwitch(diffuser))
+    for hublot, diffuser in diffusers.items():
+        coordinator = coordinators[hublot]
+        entities.append(DiffuserSwitch(diffuser, coordinator))
 
-    async_add_entities(entities, True)
+    async_add_entities(entities)
 
 
-class DiffuserSwitch(SwitchEntity):
+class DiffuserSwitch(SwitchEntity, DiffuserEntity):
     """Representation of a diffuser switch."""
 
-    def __init__(self, diffuser):
-        """Initialize the switch."""
-        self._diffuser = diffuser
-        self._available = True
-
-    @property
-    def device_info(self):
-        """Return information about the device."""
-        return {
-            "name": self._diffuser.data["hub"]["attributes"]["roomnamec"],
-            "identifiers": {(DOMAIN, self._diffuser.data["hub"]["hublot"])},
-            "manufacturer": MANUFACTURER,
-            "model": MODEL,
-            "sw_version": self._diffuser.data["hub"]["sensors"]["versionc"],
-        }
-
-    @property
-    def unique_id(self):
-        """Return the unique ID of the device."""
-        return self._diffuser.data["hub"]["hublot"]
+    def __init__(self, diffuser, coordinator):
+        """Initialize the diffuser switch."""
+        super().__init__(diffuser, coordinator, "")
+        self._is_on = self.coordinator.data[HUB][ATTRIBUTES][FAN] == ON_STATE
 
     @property
     def available(self):
         """Return if the device is available."""
-        return self._available
-
-    @property
-    def name(self):
-        """Return the name of the device."""
-        return self._diffuser.data["hub"]["attributes"]["roomnamec"]
+        return self.coordinator.data[HUB][STATUS] == AVAILABLE_STATE
 
     @property
     def icon(self):
         """Return the icon of the device."""
-        return ICON
+        return "mdi:fan"
 
     @property
     def extra_state_attributes(self):
         """Return the device state attributes."""
         attributes = {
-            "fan_speed": self._diffuser.data["hub"]["attributes"]["speedc"],
-            "room_size": self._diffuser.data["hub"]["attributes"]["roomc"],
+            "fan_speed": self.coordinator.data[HUB][ATTRIBUTES][SPEED],
+            "room_size": self.coordinator.data[HUB][ATTRIBUTES][ROOM],
```

Comment:
This should also probably be a separate sensor entity.

Reply:
`fan_speed` and `room_size` are settings in the app.
I will add a number platform for `fan_speed` and `room_size` so you can change them.
https://github.com/home-assistant/core/issues/47321#issuecomment-812887265

Reply:
Ok. Please remove them here then.